File: tools/xgmtool/src/gd3.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../inc/gd3.h"
#include "../inc/xgmtool.h"
#include "../inc/util.h"


const char* XGMTOOL_PRINT = "Optimized with XGMTool";


static int getWideStringSize(void* in);
static short* getWideStringFromString(char* str);
static short* getWideString(unsigned char* in, int *outSize);
static char* getString(short* wstr);


GD3* GD3_create()
{
    GD3* result;

    result = malloc(sizeof(GD3));
    memset(result, 0, sizeof(GD3));

    result->version = 0x100;

    return result;
}

XD3* XD3_create()
{
    XD3* result;

    result = malloc(sizeof(XD3));
    memset(result, 0, sizeof(XD3));

    return result;
}

GD3* GD3_createFromData(unsigned char* data)
{
    GD3* result = GD3_create();
    int offset = 0;
    int size;

    if (!silent)
        printf("Parsing GD3...\n");

    if (strncasecmp(data + offset, "Gd3 ", 4))
    {
        printf("Error: GD3 header not recognized !\n");
        return NULL;
    }
    offset += 4;

    // get version info
    result->version = getInt(data, 4);
    offset += 4;

    // get total size for GD3 data (we can ignore this field)
    size = getInt(data, 8);
    offset += 4;

    result->trackName_EN = getWideString(data + offset, &size);
    offset += size * 2;
    result->trackName_JP = getWideString(data + offset, &size);
    offset += size * 2;
    result->gameName_EN = getWideString(data + offset, &size);
    offset += size * 2;
    result->gameName_JP = getWideString(data + offset, &size);
    offset += size * 2;
    result->systemName_EN = getWideString(data + offset, &size);
    offset += size * 2;
    result->systemName_JP = getWideString(data + offset, &size);
    offset += size * 2;
    result->authorName_EN = getWideString(data + offset, &size);
    offset += size * 2;
    result->authorName_JP = getWideString(data + offset, &size);
    offset += size * 2;
    result->date = getWideString(data + offset, &size);
    offset += size * 2;
    result->vgmConversionAuthor = getWideString(data + offset, &size);
    offset += size * 2;
    result->notes = getWideString(data + offset, &size);

    // convert to classic string
    char* str = getString(result->vgmConversionAuthor);
    // not ending with our signature ?
    if ((strlen(str) < strlen(XGMTOOL_PRINT)) || strncmp(str + strlen(str) - strlen(XGMTOOL_PRINT), XGMTOOL_PRINT, strlen(XGMTOOL_PRINT)))
    {
        // realloc string
        str = realloc(str, strlen(str) + 4 + strlen(XGMTOOL_PRINT));
        // print XGMTool signature
        strcat(str, " - ");
        strcat(str, XGMTOOL_PRINT);
    }
    // convert back to wide string
    result->vgmConversionAuthor = getWideStringFromString(str);

    return result;
}

XD3* XD3_createFromGD3(GD3 *gd3, int duration, int loopDuration)
{
    XD3* result = XD3_create();

    if (!silent)
        printf("Converting GD3 to XD3...\n");

    result->trackName = getString(gd3->trackName_EN);
    result->gameName = getString(gd3->gameName_EN);
    result->authorName = getString(gd3->authorName_EN);
    result->date = getString(gd3->date);
    result->conversionAuthor = getString(gd3->vgmConversionAuthor);
    result->notes = getString(gd3->notes);
    result->duration = duration;
    result->loopDuration = loopDuration;

    return result;
}

int GD3_computeDataSize(GD3* gd3)
{
    return (getWideStringSize(gd3->trackName_EN) * 2) +
        (getWideStringSize(gd3->trackName_JP) * 2) +
        (getWideStringSize(gd3->gameName_EN) * 2) +
        (getWideStringSize(gd3->gameName_JP) * 2) +
        (getWideStringSize(gd3->systemName_EN) * 2) +
        (getWideStringSize(gd3->systemName_JP) * 2) +
        (getWideStringSize(gd3->authorName_EN) * 2) +
        (getWideStringSize(gd3->authorName_JP) * 2) +
        (getWideStringSize(gd3->date) * 2) +
        (getWideStringSize(gd3->vgmConversionAuthor) * 2) +
        (getWideStringSize(gd3->notes) * 2) + (11 * 2);
}

int XD3_computeDataSize(XD3* xd3)
{
    return strlen(xd3->trackName) +
        strlen(xd3->gameName) +
        strlen(xd3->authorName) +
        strlen(xd3->date) +
        strlen(xd3->conversionAuthor) +
        strlen(xd3->notes) + (6 * 1) + 8;   // +8 for durations informations
}


unsigned char* GD3_asByteArray(GD3* gd3, int *outSize)
{
    int size = GD3_computeDataSize(gd3);
    unsigned char* result = malloc(size + 12);
    int offset = 0;

    // header
    strcpy(result, "Gd3 ");
    offset += 4;
    // version
    setInt(result, offset, gd3->version);
    offset += 4;
    // size
    setInt(result, offset, size);
    offset += 4;

    // set out size
    *outSize = size + 12;

    // fields
    size = (getWideStringSize(gd3->trackName_EN) * 2) + 2;
    memcpy(result + offset, gd3->trackName_EN, size);
    offset += size;
    size = (getWideStringSize(gd3->trackName_JP) * 2) + 2;
    memcpy(result + offset, gd3->trackName_JP, size);
    offset += size;
    size = (getWideStringSize(gd3->gameName_EN) * 2) + 2;
    memcpy(result + offset, gd3->gameName_EN, size);
    offset += size;
    size = (getWideStringSize(gd3->gameName_JP) * 2) + 2;
    memcpy(result + offset, gd3->gameName_JP, size);
    offset += size;
    size = (getWideStringSize(gd3->systemName_EN) * 2) + 2;
    memcpy(result + offset, gd3->systemName_EN, size);
    offset += size;
    size = (getWideStringSize(gd3->systemName_JP) * 2) + 2;
    memcpy(result + offset, gd3->systemName_JP, size);
    offset += size;
    size = (getWideStringSize(gd3->authorName_EN) * 2) + 2;
    memcpy(result + offset, gd3->authorName_EN, size);
    offset += size;
    size = (getWideStringSize(gd3->authorName_JP) * 2) + 2;
    memcpy(result + offset, gd3->authorName_JP, size);
    offset += size;
    size = (getWideStringSize(gd3->date) * 2) + 2;
    memcpy(result + offset, gd3->date, size);
    offset += size;
    size = (getWideStringSize(gd3->vgmConversionAuthor) * 2) + 2;
    memcpy(result + offset, gd3->vgmConversionAuthor, size);
    offset += size;
    size = (getWideStringSize(gd3->notes) * 2) + 2;
    memcpy(result + offset, gd3->notes, size);

    return result;
}

unsigned char* XD3_asByteArray(XD3* xd3, int *outSize)
{
    // align size on 2 bytes
    int size = (XD3_computeDataSize(xd3) + 1) & 0xFFFFFFFE;
    unsigned char* result = malloc(size + 4);
    int offset = 0;

    // size of XD3 infos
    setInt(result, offset, size);
    offset += 4;

    // set out size
    *outSize = size + 4;

    // fields
    size = strlen(xd3->trackName) + 1;
    memcpy(result + offset, xd3->trackName, size);
    offset += size;
    size = strlen(xd3->gameName) + 1;
    memcpy(result + offset, xd3->gameName, size);
    offset += size;
    size = strlen(xd3->authorName) + 1;
    memcpy(result + offset, xd3->authorName, size);
    offset += size;
    size = strlen(xd3->date) + 1;
    memcpy(result + offset, xd3->date, size);
    offset += size;
    size = strlen(xd3->conversionAuthor) + 1;
    memcpy(result + offset, xd3->conversionAuthor, size);
    offset += size;
    size = strlen(xd3->notes) + 1;
    memcpy(result + offset, xd3->notes, size);
    offset += size;
    setInt(result, offset + 0, xd3->duration);
    setInt(result, offset + 4, xd3->loopDuration);

    return result;
}


static int getWideStringSize(void* in)
{
    short* src = (short*) in;
    int result = 0;

    while(*src++) result++;

    return result;
}


static short* getWideStringFromString(char* str)
{
    int size = strlen(str) + 1;
    short* result = malloc(size * sizeof(short));
    int i;

    // set to 0 first
    memset(result, 0, size * sizeof(short));
    // then convert
    for(i = 0; i < size; i++)
        result[i] = str[i];

    return result;
}

static short* getWideString(unsigned char* in, int *outSize)
{
    int size = getWideStringSize(in) + 1;
    short* result = malloc(size * sizeof(short));

    memcpy(result, in, size * sizeof(short));

    *outSize = size;

    return result;
}

static char* getString(short* wstr)
{
    int size = getWideStringSize(wstr) + 1;
    char* result = malloc(size * sizeof(char));
    int i;

    // convert to byte string
    for(i = 0; i < size; i++)
        result[i] = wstr[i];

    return result;
}

File: tools/crates/xgmtool/src/gd3.rs
/// GD3 tags (VGM metadata, uses UTF-16LE wide strings)
#[derive(Clone, Debug)]
pub struct GD3 {
    pub version: u32,
    pub track_name_en: Vec<u16>,
    pub track_name_jp: Vec<u16>,
    pub game_name_en: Vec<u16>,
    pub game_name_jp: Vec<u16>,
    pub system_name_en: Vec<u16>,
    pub system_name_jp: Vec<u16>,
    pub author_name_en: Vec<u16>,
    pub author_name_jp: Vec<u16>,
    pub date: Vec<u16>,
    pub vgm_conversion_author: Vec<u16>,
    pub notes: Vec<u16>,
}

/// XD3 tags (simplified XGM metadata, ASCII strings)
#[derive(Clone, Debug)]
pub struct XD3 {
    pub track_name: String,
    pub game_name: String,
    pub author_name: String,
    pub date: String,
    pub conversion_author: String,
    pub notes: String,
    pub duration: i32,
    pub loop_duration: i32,
}

const XGMTOOL_PRINT: &str = "Optimized with XGMTool";

impl GD3 {
    pub fn new() -> Self {
        GD3 {
            version: 0x100,
            track_name_en: vec![0], track_name_jp: vec![0],
            game_name_en: vec![0], game_name_jp: vec![0],
            system_name_en: vec![0], system_name_jp: vec![0],
            author_name_en: vec![0], author_name_jp: vec![0],
            date: vec![0],
            vgm_conversion_author: vec![0],
            notes: vec![0],
        }
    }

    pub fn from_data(data: &[u8]) -> Option<Self> {
        if !crate::is_silent() {
            println!("Parsing GD3...");
        }

        if data.len() < 12 {
            println!("Error: GD3 data too small!");
            return None;
        }

        let header = std::str::from_utf8(&data[0..4]).unwrap_or("");
        if !header.eq_ignore_ascii_case("Gd3 ") {
            println!("Error: GD3 header not recognized !");
            return None;
        }

        let mut gd3 = GD3::new();
        gd3.version = crate::get_u32(data, 4);
        // size at offset 8 (ignored)
        let mut offset = 12;

        gd3.track_name_en = read_wide_string(data, &mut offset);
        gd3.track_name_jp = read_wide_string(data, &mut offset);
        gd3.game_name_en = read_wide_string(data, &mut offset);
        gd3.game_name_jp = read_wide_string(data, &mut offset);
        gd3.system_name_en = read_wide_string(data, &mut offset);
        gd3.system_name_jp = read_wide_string(data, &mut offset);
        gd3.author_name_en = read_wide_string(data, &mut offset);
        gd3.author_name_jp = read_wide_string(data, &mut offset);
        gd3.date = read_wide_string(data, &mut offset);
        gd3.vgm_conversion_author = read_wide_string(data, &mut offset);
        gd3.notes = read_wide_string(data, &mut offset);

        // Append XGMTool signature to conversion author
        let author_str = wide_to_string(&gd3.vgm_conversion_author);
        if !author_str.ends_with(XGMTOOL_PRINT) {
            let new_author = format!("{} - {}", author_str, XGMTOOL_PRINT);
            gd3.vgm_conversion_author = string_to_wide(&new_author);
        }

        Some(gd3)
    }

    pub fn compute_data_size(&self) -> usize {
        (self.track_name_en.len() + self.track_name_jp.len() +
         self.game_name_en.len() + self.game_name_jp.len() +
         self.system_name_en.len() + self.system_name_jp.len() +
         self.author_name_en.len() + self.author_name_jp.len() +
         self.date.len() + self.vgm_conversion_author.len() +
         self.notes.len()) * 2
    }

    pub fn as_byte_array(&self) -> Vec<u8> {
        let data_size = self.compute_data_size();
        let mut result = Vec::with_capacity(12 + data_size);

        // Header
        result.extend_from_slice(b"Gd3 ");
        result.extend_from_slice(&self.version.to_le_bytes());
        result.extend_from_slice(&(data_size as u32).to_le_bytes());

        // Fields
        write_wide_string(&mut result, &self.track_name_en);
        write_wide_string(&mut result, &self.track_name_jp);
        write_wide_string(&mut result, &self.game_name_en);
        write_wide_string(&mut result, &self.game_name_jp);
        write_wide_string(&mut result, &self.system_name_en);
        write_wide_string(&mut result, &self.system_name_jp);
        write_wide_string(&mut result, &self.author_name_en);
        write_wide_string(&mut result, &self.author_name_jp);
        write_wide_string(&mut result, &self.date);
        write_wide_string(&mut result, &self.vgm_conversion_author);
        write_wide_string(&mut result, &self.notes);

        result
    }
}

impl XD3 {
    pub fn new() -> Self {
        XD3 {
            track_name: String::new(),
            game_name: String::new(),
            author_name: String::new(),
            date: String::new(),
            conversion_author: String::new(),
            notes: String::new(),
            duration: 0,
            loop_duration: 0,
        }
    }

    pub fn from_gd3(gd3: &GD3, duration: i32, loop_duration: i32) -> Self {
        if !crate::is_silent() {
            println!("Converting GD3 to XD3...");
        }
        XD3 {
            track_name: wide_to_string(&gd3.track_name_en),
            game_name: wide_to_string(&gd3.game_name_en),
            author_name: wide_to_string(&gd3.author_name_en),
            date: wide_to_string(&gd3.date),
            conversion_author: wide_to_string(&gd3.vgm_conversion_author),
            notes: wide_to_string(&gd3.notes),
            duration,
            loop_duration,
        }
    }

    pub fn compute_data_size(&self) -> usize {
        self.track_name.len() + self.game_name.len() + self.author_name.len()
            + self.date.len() + self.conversion_author.len() + self.notes.len()
            + 6 + 8 // 6 null terminators + 8 bytes for durations
    }

    pub fn as_byte_array(&self) -> Vec<u8> {
        let size = (self.compute_data_size() + 1) & !1; // align to 2 bytes
        let mut result = Vec::with_capacity(size + 4);

        result.extend_from_slice(&(size as u32).to_le_bytes());

        write_c_string(&mut result, &self.track_name);
        write_c_string(&mut result, &self.game_name);
        write_c_string(&mut result, &self.author_name);
        write_c_string(&mut result, &self.date);
        write_c_string(&mut result, &self.conversion_author);
        write_c_string(&mut result, &self.notes);
        result.extend_from_slice(&(self.duration as u32).to_le_bytes());
        result.extend_from_slice(&(self.loop_duration as u32).to_le_bytes());

        // Pad to target size
        while result.len() < size + 4 {
            result.push(0);
        }

        result
    }
}

// --- Helper functions ---

fn read_wide_string(data: &[u8], offset: &mut usize) -> Vec<u16> {
    let mut result = Vec::new();
    while *offset + 1 < data.len() {
        let ch = (data[*offset] as u16) | ((data[*offset + 1] as u16) << 8);
        *offset += 2;
        result.push(ch);
        if ch == 0 { break; }
    }
    if result.is_empty() || *result.last().unwrap() != 0 {
        result.push(0);
    }
    result
}

fn write_wide_string(buf: &mut Vec<u8>, s: &[u16]) {
    for &ch in s {
        buf.push(ch as u8);
        buf.push((ch >> 8) as u8);
    }
}

fn wide_to_string(wstr: &[u16]) -> String {
    wstr.iter()
        .take_while(|&&c| c != 0)
        .map(|&c| c as u8 as char)
        .collect()
}

fn string_to_wide(s: &str) -> Vec<u16> {
    let mut result: Vec<u16> = s.bytes().map(|b| b as u16).collect();
    result.push(0);
    result
}

fn write_c_string(buf: &mut Vec<u8>, s: &str) {
    buf.extend_from_slice(s.as_bytes());
    buf.push(0);
}
